### ISSUE TITLE
Feature Request: Group/Device Rates Output

### ISSUE THREAD

**Issue by KadarMarine** [nicehash/NiceHashMiner#1671]
In the older versions (v1.9.0.23) it was a relatively easy task for my Rig Monitoring Software (RigMinder) to read the window displays for Group/Device Rates as displayed below...

![image](https://user-images.githubusercontent.com/7789961/61175747-aee3b180-a582-11e9-9198-4a09be42dd7b.png)

In the new version I find it impossible to gain access to the data shown in the new display...
![image](https://user-images.githubusercontent.com/7789961/61175774-04b85980-a583-11e9-9d22-625a0d01b332.png)

My Monitoring Software is an integral part of my mining operation. It sends this data to a Server over the LAN (once per second). The server displays each Mining Rig in it's own display window on my Base PC; and notifies me in real time if there is a problem with the mining rig...
![image](https://user-images.githubusercontent.com/7789961/61175833-249c4d00-a584-11e9-900a-ffd808824049.png)
This software is capable of monitoring up to 63 Rig Clients, and also provides shutdown/startup automation for the individual mining rig clients based on profitability.

Since this data is no longer accessible in this new version of NiceHash Miner; could you add a feature that would allow this data to be piped (in real time) to an external program?

My preference would be a console utility (.exe) that could be started by my software and run as a pipe thread whenever monitoring begins. I could read the displayed data through the pipe thread. Those not using such monitoring software could simply ignore this additional utility. This console feature could use the CreateFileMapping and MapViewOfFile functions to obtain the data from NiceHash Miner whenever it is started through my client monitoring program. The console should update once per second by default, or the update timing could be specified through the command line.

If on the other hand there already exists a means of gathering this data from NiceHash Miner, I'd appreciate a hint or two to point the way forward.

Thanks!

For more insight on the RigMinder software for the old NiceHash Miner Legacy platform , the NhlRigMinder software can be downloaded here: 
https://kadarmarine-ftp.sharefile.com/d-sdbf8bd0130c4decb
Documentation files are included. My contact info is given on the last page of each help file.
Requires a PC running Windows.

**Comment by S74nk0** [comment]
Hi @KadarMarine

> If on the other hand there already exists a means of gathering this data from NiceHash Miner, I'd appreciate a hint or two to point the way forward.

this currently doesn't exist but I was thinking that such a feature might being useful myself. We can work together on this feature just give me what kind of data format you need or we can work out a generic data model and pipe it out to some external utility that transforms the data further (giving more leverage to external users).

For simplicity we can add a JSON based data model and maybe some other format in the future. 

Now as for 1second interval data updates this is not impossible but currently the miner stats API calls are the same interval for all miners. We can separate this with the internal settings if this would be beneficial to this feature.

I will check out the documentation for you provided later and we can continue with solving this issue.

**Comment by KadarMarine** [comment]
Hi Stanko,

The JSON format would be easy to parse into data packets for transmission over the LAN connection.

Since there was previously no trigger available to base the timing on; my software defaulted to a 1 second interval to provide timely updates to the monitoring PC. I note that the current display in the new NiceHash Miner updates at a rate of about 4-5 seconds. If the new external utility displayed at the same rate, that would be perfect. I would send the LAN transmission packets whenever new data is received through the pipe.

The inclusion of the Device Status information (Temperature, Power, etc) is also beneficial to my effort. I currently use the nvidia-smi.exe utility to gather this information. Over time I realize that this utility sometimes "fades", loosing touch with some of the devices it is reporting on. I will be able to scrap that utility pipe thread; and provide the same data for the other device types when available. 

Many thanks for your consideration in this. Let me know if there are any questions or other concerns.

**Comment by KadarMarine** [comment]
Hi Stanko,

I'm currently working on a "work around" to get my automation system working again.

Currently my only option is to have my RigMinder client program start NiceHash Miner and run it under a pipe thread in order to receive the stdout (logs).

In doing so I have established a list of possible outputs that would be preferable for the console utility (enhancement) that we discussed....

The sma data appears to update in a timely enough manner (about once a minute):
...13:40:49] [INFO] [SOCKET] Received: {"method":"sma","data":[[14,"7.560593087e-05"],...

The ExchangeRateApi updates about every 15 minutes, once a minute would be preferable:
...12:26:13] [INFO] [ExchangeRateApi] USD rate updated: 8313.311523 BTC

The Current global profit (when MinProfit is in use) appears to update in a timely enough manner:
...] [INFO] [MiningManager] Current global profit = 3.08606992 USD/Day
...] [INFO] [MiningManager] Current global profit = IS PROFITABLE, MinProfit: 3.00000000 USD/Day

The CurrentProfit is kind of nice, but somewhat unnecessary since the RigMinder is calculating this: 
... [MiningManager] PrevStateProfit 0.000384478023616583, CurrentProfit 0.000384478023616583

The "miner.status" PENDING, STOPPED, BENCHMARKING, and MINING outputs contain all the other information I need, but I only get these about every 45 seconds. I understand why you don't want to send this over the SOCKET more frequently than that; but it would be great if the console utility enhancement could spit out an update for "miner.status" info about every 5 seconds.... 
[SOCKET] Sending data: {"method":"miner.status","params":["MINING",[["","2-pQG5wU0ctVO2HkExsK6mmA",18,99,[[46,23904183.6]],67,3944,123,-1,0],...

I use the "miner.status" "PENDING" output to gather the CPU/GPU names; it would be great if we could output a similar "miner.status" line which contained the CPU/GPU names soon after the console utility enhancement is initialized (under my pipe thread).

Additional Note:
I would also like to see the following two entries in the console utility enhancement:
[2019-10-31 12:15:27] [INFO] [MiningManager] Start New Mining : (...
[2019-10-31 12:15:27] [INFO] [MiningManager] Stop Old Mining: (...
With these I will be able track which miners are currently deployed.   

Thanks again.
